presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java

```diff
@@ -334,37 +344,86 @@ public int getIndex()
+                    checkArgument(add(projectionsForColumn, column.getBaseHiveColumnIndex(), column.getHiveColumnProjectionInfo()), "duplicate column in columns list");
+
+                    // Add regular mapping if projection is valid for partition schema, otherwise add an empty mapping
+                    if (projectionExistsInPartitionSchema(baseTypeCoercionFrom, column)) {
+                        columnMappings.add(regular(column, regularIndex, baseTypeCoercionFrom));
+                        regularIndex++;
+                    }
+                    else {
+                        columnMappings.add(empty(column));
+                    }
                 }
                 else {
                     columnMappings.add(prefilled(
                             column,
                             getPrefilledColumnValue(column, partitionKeysByName.get(column.getName()), path, bucketNumber, fileSize, fileModifiedTime),
-                            coercionFrom));
+                            baseTypeCoercionFrom));
                 }
             }
+
             for (HiveColumnHandle column : requiredInterimColumns) {
                 checkArgument(column.getColumnType() == REGULAR);
-                if (regularColumnIndices.contains(column.getHiveColumnIndex())) {
+                checkArgument(column.isBaseColumn(), "bucketed columns should be base columns");
+                if (baseColumnHiveIndices.contains(column.getBaseHiveColumnIndex())) {
                     continue; // This column exists in columns. Do not add it again.
                 }
-                // If coercion does not affect bucket number calculation, coercion doesn't need to be applied here.
-                // Otherwise, read of this partition should not be allowed.
-                // (Alternatively, the partition could be read as an unbucketed partition. This is not implemented.)
-                columnMappings.add(interim(column, regularIndex));
+
+                if (projectionsForColumn.containsKey(column.getBaseHiveColumnIndex())) {
+                    columnMappings.add(interim(column, regularIndex, Optional.ofNullable(columnCoercions.get(column.getBaseHiveColumnIndex()))));
+                }
+                else {
+                    // If coercion does not affect bucket number calculation, coercion doesn't need to be applied here.
+                    // Otherwise, read of this partition should not be allowed.
+                    // (Alternatively, the partition could be read as an unbucketed partition. This is not implemented.)
+                    columnMappings.add(interim(column, regularIndex, Optional.empty()));
+                }
                 regularIndex++;
             }
             return columnMappings.build();
         }
 
+        private static boolean projectionExistsInPartitionSchema(Optional<HiveType> baseTypeCoercionFrom, HiveColumnHandle column)
+        {
+            if (!baseTypeCoercionFrom.isPresent()) {
```

Comment:
I see what you mean. I've taken the `baseTypeCoercionFrom.isPresent()` check out since it's the semantics of the caller method. also renamed the method to `projectionExistsForType`.